src/Shape.cpp

```diff
@@ -25,7 +26,30 @@ const Mesh& Shape::getMesh() const {
     return mesh_;
 }
 
+const Box Shape::getBoundingBox() const {
+    if (!bounding_box_cache_valid_)
+    {
+        const std::vector<geo::Vector3>& points = mesh_.getPoints();
+        double x_min, y_min, z_min = 1e9;
+        double x_max, y_max, z_max = -1e9;
+        for (std::vector<geo::Vector3>::const_iterator it = points.begin(); it != points.end(); ++it)
+        {
+            x_min = std::min<double>(it->x, x_min);
+            x_max = std::max<double>(it->x, x_max);
+            y_min = std::min<double>(it->y, y_min);
+            y_max = std::max<double>(it->y, y_max);
+            z_min = std::min<double>(it->z, z_min);
+            z_max = std::max<double>(it->z, z_max);
+        }
+        bounding_box_min_cache_ = geo::Vector3(x_min, y_min, z_min);
+        bounding_box_max_cache_ = geo::Vector3(x_max, y_max, z_max);
+        bounding_box_cache_valid_ = true;
+    }
+    return Box(bounding_box_min_cache_, bounding_box_max_cache_);
+}
+
 void Shape::setMesh(const Mesh& mesh) {
+    bounding_box_cache_valid_ = false;
```

Comment:
Such a nice consistency problem here.

There is a shape holding one or more child shapes. Assume one of these childs is this object. I query the bounding box of the composite shape, then call this method on one of the childs, changing its content.

Now you have a composite shape that is not telling the truth about its bounding box.

Reply:
Interesting, I will think about a solution.

Reply:
I somewhat wonder why we don't use a well developed and maintained library for this instead. It's not like a hierarchy of shapes is anywhere near a ground breaking novel data structure, is it?

Reply:
This current implementation also has a consistency problem. Because the bounding box of `Composite` is set in `addshape`. Because the childs are stored as a `boost::shared_ptr`. This could be edited from outside this class. A simple solution would be to store const pointers. Do you agree?

Reply:
Correction: in `addshape` the object in the shared pointer is cloned. So it can't be edited from outside.